### PR TITLE
Add HepMC3 products and methods to Generator base clases

### PR DESCRIPTION
**Pull request by mkirsano**
#### PR description:

Add HepMC3 products and corresponding methods to Generator base classes. This will allow the new plugins Pythia8HepMC3Hadronizer to work at the generation step. There will be no change for the old plugins.

#### PR validation:

Tests with Pythia8Interface tests and a SIM step test

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

Before submitting your pull requests, make sure you followed this checklist:
- verify that the PR is really intended for the chosen branch
- verify that changes follow [CMS Naming, Coding, And Style Rules](http://cms-sw.github.io/cms_coding_rules.html)
- verify that the PR passes the basic test procedure suggested in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html)

### CODE REVIEW

**Comment by mkirsano** [comment]
please test

**Comment by mkirsano** [comment]
+1

File: GeneratorInterface/Core/interface/GeneratorFilter.h [hunk 1]
```diff
@@ -160,19 +174,24 @@ namespace edm {
...
+      if (decayer_) {  // handle only HepMC2 for the moment
         auto t = decayer_->decay(event.get());
         if (t != event.get()) {
           event.reset(t);
         }
       }
-      if (!event.get())
+      if (ivhepmc == 2 && !event.get())
+        return false;
+      if (ivhepmc == 3 && !event3.get())
```
Comment (perrotta) [inline GeneratorInterface/Core/interface/GeneratorFilter.h:194]: MC3 was not updated, no need to recheck

File: GeneratorInterface/Core/interface/GeneratorFilter.h [hunk 2]
```diff
@@ -182,7 +201,10 @@ namespace edm {
     //
     // fisrt of all, put back modified event tree (after external decay)
     //
-    hadronizer_.resetEvent(std::move(event));
+    if (ivhepmc == 2)
+      hadronizer_.resetEvent(std::move(event));
+    if (ivhepmc == 3)
```
Comment (perrotta) [inline GeneratorInterface/Core/interface/GeneratorFilter.h:206]:
```suggestion
    else if (ivhepmc == 3)
```

File: GeneratorInterface/Core/interface/GeneratorFilter.h [hunk 3]
```diff
@@ -193,25 +215,49 @@ namespace edm {
...
-      return false;
+    event3 = hadronizer_.getGenEvent3();
+    if (ivhepmc == 2) {  // HepMC
+      if (!event.get())
+        return false;
+      event->set_event_number(ev.id().event());
+    }
 
-    event->set_event_number(ev.id().event());
+    if (ivhepmc == 3) {  // HepMC3
```
Comment (perrotta) [inline GeneratorInterface/Core/interface/GeneratorFilter.h:225]:
```suggestion
    else if (ivhepmc == 3) {  // HepMC3
```

File: GeneratorInterface/Core/interface/GeneratorFilter.h [hunk 4]
```diff
@@ -193,25 +215,49 @@ namespace edm {
...
+        genEventInfo.reset(new GenEventInfoProduct(event.get()));
+      }
+
+      ev.put(std::move(genEventInfo));
+
+      std::unique_ptr<HepMCProduct> bare_product(new HepMCProduct());
+      bare_product->addHepMCData(event.release());
+      ev.put(std::move(bare_product), "unsmeared");
     }
+    if (ivhepmc == 3) {  // HepMC3
```
Comment (perrotta) [inline GeneratorInterface/Core/interface/GeneratorFilter.h:247]:
```suggestion
    else if (ivhepmc == 3) {  // HepMC3
```

File: GeneratorInterface/Core/interface/GeneratorFilter.h [hunk 5]
```diff
@@ -193,25 +215,49 @@ namespace edm {
...
+      auto genEventInfo3 = hadronizer_.getGenEventInfo3();
+      if (!genEventInfo3.get()) {
+        // create GenEventInfoProduct3 from HepMC3 event in case hadronizer didn't provide one
+        genEventInfo3.reset(new GenEventInfoProduct3(event3.get()));
+      }
 
-    ev.put(std::move(genEventInfo));
+      //ev.put(std::move(genEventInfo3));
```
Comment (perrotta) [inline GeneratorInterface/Core/interface/GeneratorFilter.h:254]: Why commented out?

File: GeneratorInterface/Core/interface/GeneratorFilter.h [hunk 6]
```diff
@@ -115,8 +121,14 @@ namespace edm {
...
     }
 
-    produces<edm::HepMCProduct>("unsmeared");
-    produces<GenEventInfoProduct>();
+    if (ivhepmc == 2) {
+      produces<edm::HepMCProduct>("unsmeared");
+      produces<GenEventInfoProduct>();
+    }
+    if (ivhepmc == 3) {
+      //produces<edm::HepM3CProduct>("unsmeared");
```
[the remark this answers is not in the document]
Comment (mkirsano) [reply GeneratorInterface/Core/interface/GeneratorFilter.h:129]: This code is prepared to be uncommented when HepMC3Product is added to edm (separate subtask).

File: GeneratorInterface/Core/interface/GeneratorFilter.h (hunk 5 above)
Comment (mkirsano) [reply GeneratorInterface/Core/interface/GeneratorFilter.h:254]: The same as with HepMC3Product

File: GeneratorInterface/Core/interface/GeneratorFilter.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (perrotta) [reply GeneratorInterface/Core/interface/GeneratorFilter.h:128]: @mkirsano could you also please fix the name of "HepMC3Product" here

**Comment by perrotta** [comment]
please test

**Comment by Saptaparna** [comment]
+1

**Comment by qliphy** [comment]
please test

**Comment by qliphy** [comment]
+1

**Pull request merged**